### PR TITLE
Authorization at article level in com_content not always working: Featured and Articles models

### PR DESCRIPTION
**Pull request by LivioCavallo**
### Summary of Changes
Joomla! has a fantastic permissions system that allows fine-grained settings in components, categories till to the article level.
In the major part of cases all these settings are correctly checked and respected but in some circumstancies 'com_content' only checks permissions at component level: article, articles and featured models.

------------------------------------------------------------------------------------------------------------
_[The following problem is now in PR #17735 - BEGIN]_
1. article model:
- `populateState` function authorises at component level and eventually filters published and archived articles.
A 'TODO' in comments seems to me to clearly suggest that a different authorisation was intended.
This PR authorises at article level.

- `getItem` function authorises at component level and eventually filters not-expired and not-future articles.
It seems to me that article level authorisation was simply forgotten.
This PR authorises at article level.

_[Previous problem is now in PR #17735 - END]_

------------------------------------------------------------------------------------------------------------

2. articles model:
- `populateState` function authorises at component level and eventually filters only published articles.
In fact I think that it cannot do otherwisely as we still do not have article ids here.
The only way to authorise at article level here would be to build a SQL query to do all the authorization work, but I think it's impossible, given the complexity of `authorise` code.
This PR postpones authorization and article selection, based on publishing state, to `getItems` function so authorization can be done there at article level.

- `getListQuery` function authorises at component level and eventually filters not-expired and not-future articles.
Like in `populateState` function, here is the only option.
This PR analogously postpones to `getItems`.

- `getItems` function just uses `authorise` function to detect if user can edit articles.
This PR does all the previously posponed jobs: authorises, here at article level, and eventually filters out articles that are not published or are expired or future, if user hasn't 'edit' or 'edit.own' permissions.
To filter out articles that the user must not see, we remove them from the articles collection (the `$items` array) during usual `foreach` loop, so the successive code will just work on permitted/authorised articles.

3. featured model:
- `populateState` function authorises at component level and eventually filters only published articles.
Just like in articles model function this PS pospones to `getItems`.

### Testing Instructions
Create some **featured articles in different categories** (but "Uncategorised" would be enought); articles should have different publishing status: _published, unpublished, published-expired, unpublished-expired, published-in-future, unpublished-in-future_.

Create a **'Featured' menuitem**.

Set **permissions** in 'Uncategorised' category to allow 'Registered' group to 'edit' and 'edit.own'.

Create a user in registered group and login in frontend.

### Expected result
All the unpublished, expired and in-future articles in "Uncategorised" category should be displayed (as user has permissions to 'edit' and 'edit-own').

### Actual result
All the unpublished, expired and in-future articles in "Uncategorised" category are NOT displayed.

### Notes

1. I have not evaluated the performance impact of `foreach` loop in articles model `getItems` function that removes unwanted articles.

2. Should the `'filter.published'` settings be managed? How?

### CODE REVIEW

File: components/com_content/models/articles.php [hunk 1]
```diff
@@ -561,9 +556,39 @@ public function getItems()
 		// Get the global params
 		$globalParams = JComponentHelper::getParams('com_content', true);
 
+		// Array of indexes of elements to be removed from $items; to be done after this foreach loop
+		$i_to_remove=array();
```
Comment (Quy) [inline components/com_content/models/articles.php]: Add space before and after `=`

File: components/com_content/models/articles.php [hunk 2]
```diff
@@ -561,9 +556,39 @@ public function getItems()
...
+				$is_published = ($item->published == 1);
+				$is_expired = !( (($item->publish_up == $null_date) || ($item->publish_up <= $now_date)) &&
+								(($item->publish_down == $null_date) || ($item->publish_down >= $now_date)) );
+				
+				// Article to be removed
+				if (!$is_published || $is_expired)
+				{
+					/*	Annotate current key/index for later removal, for performance reasons:
+						in fact we could use directly array_splice here, but it would reset $items array index counter
+						and the foreach loop would restart from begining! */
```
Comment (Quy) [inline components/com_content/models/articles.php]: Fix spelling `begining`

File: components/com_content/models/articles.php [hunk 3]
```diff
@@ -561,9 +556,39 @@ public function getItems()
...
+			// If user has not permissions to 'edit' or 'edit.state' then keep published articles only
+			if ((!$user->authorise('core.edit.state', $asset)) || (!$user->authorise('core.edit', $asset)))
+			{								
+				$db = $this->getDbo();
+				$null_date = $db->getNullDate();
+				$now_date  = JFactory::getDate();
+				
+				$is_published = ($item->published == 1);
+				$is_expired = !( (($item->publish_up == $null_date) || ($item->publish_up <= $now_date)) &&
+								(($item->publish_down == $null_date) || ($item->publish_down >= $now_date)) );
```
Comment (Quy) [inline components/com_content/models/articles.php]: Remove tabs

File: components/com_content/models/articles.php (hunk 1 above)
Comment (Quy) [reply components/com_content/models/articles.php]: Per [Coding Standards Manual](https://developer.joomla.org/coding-standards/php-code.html#regular-variables-and-class-properties):

`Naming conventions for regular variables and class properties should be named using the "studly caps" style (also referred to as "bumpy case" or "camel caps"). The initial letter of the name is lowercase, and each letter that starts a new "word" is capitalized.`

Comment (LivioCavallo) [reply components/com_content/models/articles.php]: tks

File: components/com_content/models/articles.php (hunk 2 above)
Comment (LivioCavallo) [reply components/com_content/models/articles.php]: tks

File: components/com_content/models/articles.php (hunk 3 above)
Comment (LivioCavallo) [reply components/com_content/models/articles.php]: I think you mean row 575 and 579

**Comment by ggppdk** [comment]
Hi

1. About category view / multi-article listings 

You can not apply such changes to the loop of get items
you are removing items after the query, which has 3 problems 
- you will make pagination look broken , some pages will have 3 items some will have 10
- even if pagination is off you will show a random number of items into the category view / multi-article listings 
- you are slowing down the view when retrieving a large number of items e.g. 50 - 500 items

the decision to include or exclude published items must be part of the query

Now about the existing approach of using the 
```php
if ((!$user->authorise('core.edit.state', 'com_content')) && (!$user->authorise('core.edit', 'com_content')))
```

i agree the above is not so good either because having edit/edit.state at component level but not having edit/edit.state e.g. in a category AA , will allow someone to list articles of category AA that are not published

my approach (in our component) is always to filter out unpublished items, and use 1 extra ACL rule, only at component level to allow listing unpublished items (or if user is not guest and is owner of the items)

in any case ACL checks per item in the item loop after the query are not good option for the reasons stated above

----

2. About retrieving a single record , then yes you can apply the changes that you did
allowing viewing unpublished if user has edit / edit.state at the particular article asset is correct

you are fixing an existing approach, 

Now to comment on the existing approach

existing approach is not proper, it is not proper to do such check at DB model 

- the DB model should return the item and the controller or the view should decide to how to use it , show it or not
- it is OK if view / controller will call a method of the DB model to get view/edit/etc access, so i do not say that having ACL calculation in the DB model is bad, the bad thing is to use it inside the DB model the way that it is used

e.g. on some page visit, some plugin is triggered to do some work
and loads the article DB model to do it
now because DB model will not return the unpublished item such work can not be done
ok and the plugin has to use direct SQL queries to do the work

I argue that DB model should only have methods to calculate access view / edit / etc
but it should not make use of them

as said above my point 2 is not relevant to your PR , for number 2 you are improving the existing approach

- **my suggestion is that it should be left to the caller to decide if 
getItem() needs to check view publication state (and bypass publication state according to ACL edit/edit.state on the item) via a parameter that will be given to getItem() !
now such option does not exist

File: components/com_content/models/articles.php (hunk 3 above)
Comment (Quy) [reply components/com_content/models/articles.php]: Line 578

File: components/com_content/models/articles.php [hunk 4]
```diff
@@ -561,9 +556,39 @@ public function getItems()
...
 		{
+			// We could not authorise at article level in populateState nor in getListQuery, as it was too early: we hadn't article ids there.
+			// So we authorise here at article level, using article id.
+			$asset = 'com_content.article.' . $item->id;
+
+			// If user has not permissions to 'edit' or 'edit.state' then keep published articles only
+			if ((!$user->authorise('core.edit.state', $asset)) || (!$user->authorise('core.edit', $asset)))
+			{								
+				$db = $this->getDbo();
+				$null_date = $db->getNullDate();
```
Comment (Quy) [inline components/com_content/models/articles.php]: Please change other variables to naming convention.

**Comment by LivioCavallo** [comment]
About 1.

Slowing down: the performance penalty can came only from:
- the foreach loop at the end of getItems function that removes articles from $items array   or 
- the different authorisation: at article level than at component level.
In fact authorisations and selections are the same done at the moment by J!3.7.5.
How much do you think would be the percentual penalization?

Pagination: Oh! Pagination! Right.

About 2.
Thanks. Seems resonable.
Ok, we could move the authorisation and 'filtering' logic to view, but how to solve the pagination problem? (I think it will remain)
Or could we simply display empty placeholders and a message like "Some elements may have been hidden as you have not permissions to display them." ?

**Comment by ggppdk** [comment]
@LivioCavallo 

besides the pagination problem i also mentioned and the slow down
there is 3rd thing that i mentioned:

> even if pagination is off you will show a random number of items into the category view / multi-article listings

so i say let's not change at all the getItems method

you can make your PR just for the case of single item
with 1 modification for new records $pk is zero , do not try to check ACL on asset:
com_content.article.0

**Comment by LivioCavallo** [comment]
Do you have a measure or an estimate of the slow down?

Calling authorise on 'com_content.article.0' simply returns false when needed.
Moreover article model is not used for new records, in that case J! uses the form model, then article model will never be passed a $pk value of 0.

I cannot reproduce your problem with random items in category view / multi-article listings. When does it happen?

Yes, we now have PR #17735 for single article and this for featured and article listings.

**Comment by ggppdk** [comment]
hi, lets forget the slow down totally ! (when listing more than 50 or more than 100 items)

and focus on the **2 bugs that this PR introduces**

- you are breaking pagination, some pages will have 3 items some will have 10
- even if pagination is off you will show a random number of items into the category view / multi-article listings

Any checks of which items to include 
- must be done / decided before the SQL query items

e.g. you can fix your reported issue by another way 
suggest a PR that unpublished items will be shown to
- super-users
- item owners

or you can introduce a new ACL rule (component level only) "Ignore publication status"
the above are just some ideas

**Comment by ghost** [comment]
@LivioCavallo are the Bugs mentioned by @ggppdk are solved and this PR is good for Test?

**Comment by laoneo** [comment]
Sorry that it has taken so long to respond to this pull request. Unfortunately this pr has unknown side effects, especially with pagination. In the current state it is not something we can consider to get merged, so I'm closing it. If you still see it as an issue, then please rebase to the 4.2 branch. For the testers, please do some proper testing with front end list views and pagination. Also consider to add some system tests for API testing pagination. Thanks for your contribution and help, making Joomla better.

**Pull request closed without merging**